cmd/src/actions_exec.go

```diff
@@ -586,38 +590,116 @@ fragment repositoryFields on Repository {
 
 		// Skip repos from unsupported code hosts but don't report them explicitly.
 		if !includeUnsupported && strings.ToLower(repo.ExternalRepository.ServiceType) != "github" && strings.ToLower(repo.ExternalRepository.ServiceType) != "bitbucketserver" {
-			unsupported = append(unsupported, repo.Name)
+			unsupported[repo.Name] = struct{}{}
 			continue
 		}
 
-		if repo.DefaultBranch == nil || repo.DefaultBranch.Name == "" {
-			skipped = append(skipped, repo.Name)
-			continue
+		if _, ok := reposByID[repo.ID]; !ok {
+			reposByID[repo.ID] = &ActionRepo{
+				ID:   repo.ID,
+				Name: repo.Name,
+			}
 		}
+	}
 
-		if repo.DefaultBranch.Target.OID == "" {
-			skipped = append(skipped, repo.Name)
-			continue
+	reposQueryTemplate := `{
+  repositories(names: [%s]) {
+    nodes {
+      id
+      name
+      defaultBranch {
+        name
+        target {
+          oid
+        }
+      }
+    }
+  }
+}
+
+`
+
+	skipped := make(map[string]struct{})
+
+	// This helper function fetches a subset of repos by name and is used lower down
+	// so that we query in batches.
+	fetchRepos := func(names []string) error {
+		reposQuery := fmt.Sprintf(reposQueryTemplate, strings.Join(names, ","))
+
+		var repoResult struct {
+			Data struct {
+				Repositories struct {
+					Nodes []struct {
+						ID            string `json:"id"`
+						Name          string `json:"name"`
+						DefaultBranch *struct {
+							Name   string `json:"name"`
+							Target struct {
+								OID string `json:"oid"`
+							} `json:"target"`
+						} `json:"defaultBranch"`
+					} `json:"nodes"`
+				} `json:"repositories"`
+			} `json:"data,omitempty"`
+
+			Errors []struct {
+				Message string
+				Path    []interface{}
```

Comment:
I don't think we need them, but: you added `json:""` tags to every other field but not here.